File: 1488-avoid-flood-in-the-city/1488-avoid-flood-in-the-city.java
class Solution {
    public int[] avoidFlood(int[] rains) {
        Map<Integer,Integer>fullLakes=new HashMap<>();
        TreeSet<Integer>dry=new TreeSet<>();
        int []res=new int[rains.length];
        for(int i=0;i<rains.length;i++){
            if(rains[i]==0){
                dry.add(i);
                res[i]=1;
            }
            else{
                int lake=rains[i];
                if(fullLakes.containsKey(lake)){
                    Integer it=dry.ceiling(fullLakes.get(lake));
                    if(it==null)return new int[0];
                    res[it]=lake;
                    dry.remove(it);
                }
                res[i]=-1;
                fullLakes.put(lake,i);
            }
        }
        return res;
    }
}

File: 1488-avoid-flood-in-the-city/1488-avoid-flood-in-the-city.cpp
class Solution {
public:
    vector<int> avoidFlood(vector<int>& rains) {
        
        unordered_map<int,int>fullLakes;
        set<int>dry;
        vector<int>res;
        for(int i=0;i<rains.size();i++){
            if(rains[i]==0){
                dry.insert(i);
                res.push_back(1);
            }
            else{
                int lake=rains[i];
                if(fullLakes.find(lake)!=fullLakes.end()){
                    // auto it=lower_bound(dry.begin(),dry.end(),fullLakes[lake]);
                    auto it=dry.lower_bound(fullLakes[lake]);
                    if(it==dry.end())return {};
                    int day=*it;
                    res[day]=lake;
                    dry.erase(day);
                }
                res.push_back(-1);
                fullLakes[lake]=i;
            }
        }
        return res;
    }
};